### PR TITLE
#853 - StringMatchingRecommender does not notice when it has no training data

### LINKED ISSUE #853: StringMatchingRecommender does not notice when it has no training data
**Describe the bug**
The `StringMatchingRecommender` currently generates a sample for each sentence it encounters, no matter if that sentence actually contains annotations. This means that the check for the training/test set sizes will pass even if there is no training data at all.

**To Reproduce**
1. Import a text
2. Setup a StringMatchingRecommender to predict named entities
3. Open document and watch evaluation

**Expected behavior**
If there is no training data or too little training data, the recommender should log a corresponding message and return `0.0`.

**Please complete the following information:**
 - Version and build ID: a6f2bd5a795f0120922114b88d85d89ab96f4b9d

**Additional context**
When changing the code such that it only considers sentences with at least one annotation, the evaluation results change since samples are not only used for training but also for evaluation. Considering only sentences with annotations as samples leaves out all gold-standard sentences which have no annotations at all meaning that the recommender cannot make false-positive predictions on them. 

**Running `thatIncrementalNerEvaluationWorks` only considering sentences with at least one annotation:**
```
2019-01-12 21:01:09 TRACE StringMatchingRecommender - Extracting data took 488ms
2019-01-12 21:01:09 INFO StringMatchingRecommender - Training on [5000] items, predicting on [3687] of total [18421]
Score: 0,260814
2019-01-12 21:01:12 TRACE StringMatchingRecommender - Extracting data took 323ms
2019-01-12 21:01:12 INFO StringMatchingRecommender - Training on [10000] items, predicting on [3687] of total [18421]
Score: 0,317409
2019-01-12 21:01:14 TRACE StringMatchingRecommender - Extracting data took 244ms
2019-01-12 21:01:14 INFO StringMatchingRecommender - Training on [14734] items, predicting on [3687] of total [18421]
Score: 0,347855
```

**Running `thatIncrementalNerEvaluationWorks` only considering all sentences:**
```
2019-01-12 21:09:34 TRACE StringMatchingRecommender - Extracting data took 510ms
2019-01-12 21:09:34 INFO StringMatchingRecommender - Training on [5000] items, predicting on [6263] of total [31300]
Score: 0,219644
2019-01-12 21:09:37 TRACE StringMatchingRecommender - Extracting data took 283ms
2019-01-12 21:09:37 INFO StringMatchingRecommender - Training on [10000] items, predicting on [6263] of total [31300]
Score: 0,270006
2019-01-12 21:09:39 TRACE StringMatchingRecommender - Extracting data took 265ms
2019-01-12 21:09:39 INFO StringMatchingRecommender - Training on [15000] items, predicting on [6263] of total [31300]
Score: 0,309062
```

The `OpenNlpPosRecommender` currently considers any sentence with at least one POS annotation as a sample. So does the `OpenNlpNerRecommender`.

**The question is**:
- should we generate samples for all sentences - even those without annotations - but when checking the training set for its size consider only those samples which contain annotations?
- should we generate samples for all sentences - but then drop the unannotated samples from the training set so that learner doesn't even see them?
- ...?

### PR DESCRIPTION
**Pull request by reckart**
**What's in the PR**
- Generate samples all data, even if the samples do not contain any actual annotations
- Only consider annotations with non-empty label values when generating samples
- Skip evaluation if there is not a minimum amount of labeled data
- Improve logging messages
- Closes #853

**How to test manually**
* Create a project with a string matching recommender
* Annotate only a single word (with a feature value) for the recommender to learn and watch console log whether skipping message is logged
* Annotate more words but without a feature value and watch console log whether skipping message is logged

**Automatic testing**
* [x] PR includes unit tests

**Documentation**
* [ ] PR updates documentation

### CODE REVIEW

**Comment by reckart** [comment]
Review anybody? I'm afraid that having provided the PR, I cannot review it myself.

**Comment by reckart** [comment]
@Sibgha360 thanks for testing.
